jaxrs-api/src/main/java/jakarta/ws/rs/BadGatewayException.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2012, 2019 Oracle and/or its affiliates. All rights reserved.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0, which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ */
+
+package jakarta.ws.rs;
```

Comment:
@eclipse-ee4j/eclipse-jaxrs Do we really want to collect all kinds of error-code-exceptions as separate classes, and do we really want to have them directly in our root namespace?

Reply:
I agree that defining exceptions for all status codes is an overkill. IMO we should only add exceptions which users want to throw/catch in real world applications. I'm fine with `NotImplementedException`, but I don't think that somebody needs a `ProxyAutheticationRequiredException` in real applications.

Reply:
> I don't think that somebody needs a `ProxyAutheticationRequiredException` in real applications.

In fact I think that it would be even *wrong* to have this particular exception, as it is not an *exceptional state* to request proxy authentication.

Reply:
Well, the corresponding HTTP status code is `407`, so it is classified as a client error. However, the actually need to throw this kind of error on the server side or handle it in the client case is extremely rare. Therefore, I don't think that it is worth adding it. Some for most of the other ones.

Reply:
Agreed to all you said. I would even say that JAX-RS itself should not contain any kind of "pure sugar" classes; we should instead have a separate "official extensions" place to put those "common comfort stuff", but this is a different story.

Regarding 407: It is an *HTTP* error certainly, but it is *not* an exceptional state of the calling REST application. The application cannot know that the server expects authentication, so it sends the request. The server answers 407, so the application knows *now* that it needs to ask the user for credentials, and will request a second time. This is more like a *login handshake* than handling an actual *programming error*. Hence, no need for an *exception*, but possibly a need for some kind of new API (like a callback asking for credentials).

Reply:
> Regarding 407: It is an _HTTP_ error certainly, but it is _not_ an exceptional state of the calling REST application.

I see your point! And I think you are correct. However, I also think that 407 is a low level status code and not very often used. So I don't like the idea of added a special exception type for this.

Reply:
@jeyvison What do you think? Do you agree to remove those exceptions that are not essentially low-level status?

Reply:
copyright and `@since`  but i won't resolve this conversation 'till we decide if we keep or remove `ProxyAutheticationRequiredException` . For what you guys are saying i think we will remove it right?

Reply:
> @jeyvison What do you think? Do you agree to remove those exceptions that are not essentially low-level status?

I agree. We should keep the code clean and add more exceptions as necessary later.

Reply:
Thinking a bit more about this, i think all the exceptions like `ProxyAutheticationRequiredException` should stay. If we think exceptions as recoverable errors it makes sense for the client to send a second request with the proxy information in response to an exception (that would work as the _login handshake_ @mkarg mentioned before.  IMHO, an API for that would be an overkill.

I know that all can be verified by the response code but exceptions are more direct and really serves the purpose here.

Reply:
But this kind of handshake is the job of the JAX-RS framework, not the job of the business application. I cannot see why each application programmer should take care of that.

Reply:
For a 407, doesn't that mean that the client is unauthenticated to the proxy?  That could mean that the client needs to authenticate, but could also mean that the credentials that the client sent are incorrect.  In that case, the application would need to react accordingly - i.e. prompt the user to re-enter their password, etc.

Reply:
This is correct, but it does not imply that an exception is the right tool of chooice. Like for other parts of JAX-RS, e. g. normal (i. e. non-proxy) client authentication, this should not be part of the business application code (which an exception always would be), but part of a separate layer implemented either by an additional Client API, or by means of filters. Here is an example what I mean: http://www.adam-bien.com/roller/abien/entry/client_side_http_basic_access.

Reply:
From this example you provided, if the user provides a wrong user or password the error still needs to be treated by the developer at business level right?

Reply:
Yes, that is my understanding.

Reply:
I do not think so. If the application registers a callback for handling this case (like showing a login dialog), there still is no need for a business level *exception*. In case of 407 the callback simply would be invoked more often, just as a browser shows a login dialog unless the password is correct.

Reply:
Hmmm, so , just so i'm sure i'm not understanding it the wrong way , what you're suggesting is that we have something like `onError(...).` that the developer can use to define a callback function and then treat the _response code_ (not the exception) the way they want and if they want?

Reply:
I explicitly do *not* want anything that is called "...error..." for non-application-errors (and 407 is *not* an application error, it is just an error code *of HTTP*). JAX-RS is *not* an HTTP wrapper - the Java API for that is *Servlets* - but JAX-RS is a *REST* API. My proposal is to make up our minds what HTTP status codes are actually worth getting covered by JAX-RS as standalone exception classes, and which not. For 407 in particular, I would plea for a method to register a `onProxyLogin(lambda)` callback which provides the algorithm to fulfil in case of missing proxy credentials.

Reply:
understood. As long as we open an issue to deal with this kind of situation, i agree that we can remove some exceptions.

This is a list of exceptions that i believe that doesn't need to me commited:

- ProxyAutheticationRequiredException
- NetworkAuthenticationRequiredException
  
**maybe**:
- ExpectationFailedException
- PreconditionRequiredException

Do you folks agree?

Reply:
@eclipse-ee4j/eclipse-jaxrs Opinions?

Reply:
My personal preference:

Include:
  * ExpectationFailedException
  * NotImplementedException
  * PreconditionRequiredException
  * TooManyRequestsException

Maybe:
  * RequestedRangeNotSatisfiableException  (Should be RangeNotSatisfiableException has header was renamed in RFC)
  * RequestEntityTooLargeException
  * RequestHeaderFieldsTooLargeException
  * MethodNotAllowedException
  * LenghtRequiredException

Exclude
  * BadGatewayException
  * GatewayTimeoutException
  * ProxyAutheticationRequiredException
  * RequestUriTooLong
  * HttpVersionNotSupportedException
  * NetworkAuthenticationRequiredException
  * RequestTimeoutException

Reasoning: The exceptions in the excluded list are very technically and/or refer to proxies. The ones in the "maybe" could be useful in some situations, but only in very special cases.

Reply:
So, i agree with @chkal 's list but i would remove the  maybe list too. We can always put it back if it's needed. What you think @mkarg ?

Reply:
Looking at the RFC's description for NotImplementedException I do wonder what use case one should use that for, as this is actually what the JAX-RS engines does automatically.

Reply:
The same for PreconditionRequired.

Reply:
@mkarg I could imagine that developers would want to use `NotImplementedException` if a certain feature of the resource is not fully implemented yet and therefore throw it manually from the resource. However, I agree that it is a rare case.

Reply:
As it is a *rare* case I do not see the benefit of having it in the standard (possibly good for some "official extensions lib").

Reply:
Just removed `NotImplementedException `